cdq-analysis/pkg/devfile.go

```diff
@@ -179,63 +179,86 @@ func ValidateDevfile(log logr.Logger, URL string, token string) (shouldIgnoreDev
-							var u *url.URL
-							u, err = url.Parse(URL)
-							if err != nil {
-								log.Error(err, fmt.Sprintf("failed to parse URL from %s", URL))
-								return shouldIgnoreDevfile, nil, fmt.Errorf(fmt.Sprintf("failed to parse URL from %s", URL))
-							}
-							u.Path = path.Join(u.Path, dockerfileURI)
-							dockerfileURI = u.String()
-							_, err = CurlEndpoint(dockerfileURI)
-						}
-					}
-					if err != nil {
-						log.Error(err, fmt.Sprintf("failed to get Dockerfile from the URI %s, invalid image component: %s", URL, component.Name))
-						return shouldIgnoreDevfile, nil, fmt.Errorf(fmt.Sprintf("failed to get Dockerfile from the URI %s, invalid image component: %s", URL, component.Name))
-					}
-				}
-				if _, ok := deployCompMap[component.Name]; ok {
-					found = true
-					break
-				}
-			}
-			if !found {
-				err = fmt.Errorf("found more than one image components, but no deploy command associated with any being defined in the devfile from %s", URL)
-				log.Error(err, "failed to validate devfile")
+	} else if len(imageComp) > 1 {
+		found := false
+		for _, component := range imageComp {
+			err = validateImageComponentDockerfile(log, component, devfileLocation, token)
+			if err != nil {
+				log.Error(err, fmt.Sprintf("failed to validate the Dockerfile from the Image Component %s in the devfile", component.Name))
 				return shouldIgnoreDevfile, nil, err
 			}
+
+			if _, ok := deployCompMap[component.Name]; ok {
+				found = true
+				break
+			}
 		}
+		if !found {
+			err = fmt.Errorf("found more than one image components, but no deploy command associated with any being defined in the devfile from %s", devfileLocation)
+			log.Error(err, "failed to validate devfile")
+			return shouldIgnoreDevfile, nil, err
+		}
+	} else if len(imageComp) == 1 {
 		// TODO: if only one image component, should return a warning that no apply command being defined
+		component := imageComp[0]
+		err = validateImageComponentDockerfile(log, component, devfileLocation, token)
+		if err != nil {
+			log.Error(err, fmt.Sprintf("failed to validate the Dockerfile from the Image Component %s in the devfile", component.Name))
+			return shouldIgnoreDevfile, nil, err
+		}
 	}
 
 	return shouldIgnoreDevfile, devfileBytes, nil
 }
 
+// validateImageComponentDockerfile validates the given image component dockerfile for a devfile location
+func validateImageComponentDockerfile(log logr.Logger, component v1alpha2.Component, devfileLocation, token string) (err error) {
+	if component.Image != nil && component.Image.Dockerfile != nil && component.Image.Dockerfile.DockerfileSrc.Uri != "" {
+		dockerfileURI := component.Image.Dockerfile.DockerfileSrc.Uri
+		absoluteDockerfileURI := strings.HasPrefix(dockerfileURI, "http://") || strings.HasPrefix(dockerfileURI, "https://")
+		absoluteDevfileLocation := strings.HasPrefix(devfileLocation, "http://") || strings.HasPrefix(devfileLocation, "https://")
```

Comment:
So we have to consider:

- devfileLocation here is being updated from our controller logic ie they are converted from the source url to the raw link
- dockerfileURI is being parsed from the devfile data directly.

so if we are going to validate HTTP URLs in CDQ, it would have to be the source.git.url ie the repo URL the user passes in on the RHTAP UI..

We can add github URL validation for CDQ webhook on this user entry. Am I understanding it correctly?

Reply:
yes if this gets validated in the webhook, then we can avoid validating in the controller.  We should check and see if there's sufficient validation being done in the component controller too and open a separate issue to address this

Reply:
https://issues.redhat.com/browse/DEVHAS-532